refactor: split src/Tag.php into classes

machour opened the pull request:
As proposed in #551, this is an attempt to refactor src/Tag.php

I took care of carefully migrating the code, tag by tag and separated my commits.

I avoided changes to the maximum to keep this PR focused. It's basically the switch/case re-implemented in OOP by taking in accounts cases where there were no breaks, etc.

Unit testing seems okay, but I'm wondering if there is a better way to validate that I didn't break things. Is there a reference PDF file with all what mPDF can produce somewhere, along with the script to generate it?

machour commented:
@finwe fixed most CS error, but I'm left with this two:
*  Class name "COLUMN_BREAKTag" is not in camel caps format
*  Class name "PAGE_BREAKTag" is not in camel caps format

How would you advise renaming those two? or should we allow these two exceptions? 🤔

finwe commented:
The code should comply with PSR-1 and PSR-2 (with the exception for tabs instead of spaces), so

> Class names MUST be declared in StudlyCaps.

I am not sure about the `Tag` suffix, it may be unnecessary for concrete classes. Or are there some language conflicts with tag name?

So, the two classes you mentioned would be `Mpdf\Tag\ColumnBreak` and `Mpdf\Tag\PageBreak`.

machour commented:
@finwe the problem is with the `var` tag, which is a reserved keyword and can't be used as a class name. 
Ok for the Tag suffix in that case?

finwe commented:
OK.

machour commented:
@finwe yet another problem. mpdf supports both `<COLUMNBREAK>` and `<COLUMN_BREAK>`, which would result in `ColumnbreakTag.php` and `ColumnBreakTag.php`, which is obviously bad.

I currently implemented some kind of factory to instantiate tags:

```diff
+	/**
+	 * @param string $tag The tag name
+	 * @return \Mpdf\Tag\Tag
+	 */
+	private function getTagInstance($tag)
+	{
+		$className = 'Mpdf\Tag\\' . $tag . 'Tag';
+		if (class_exists($className)) {
+			return new $className(
+				$this->mpdf,
+				$this->cache,
+				$this->cssManager,
+				$this->form,
+				$this->otl,
+				$this->tableOfContents,
+				$this->sizeConverter,
+				$this->colorConverter,
+				$this->imageProcessor,
+				$this->languageToFont
+			);
+		}
+	}
```

do I handle this problem here? (`if ($tag === 'COLUMN_BREAK') $tag = 'COLUMNBREAK';`) and keep only one class?

machour commented:
@finwe could you get back to me on this last question? I'd like to finish this before a change happens in `Tag.php` :)

finwe commented:
Oh, sorry, postponed and then forgot.

A hardcoded map tagname <-> classname seems to me as the cleanest solution. it will also solve class names not matching the tag - such as forementioned `VarTag`.

finwe commented:
You could also have a map only for non-standard cases and when an entry does not exist in the map, you could then resort to $className = 'Mpdf\Tag\\' . camelcase($tag)

machour commented:
No worries. So you're suggesting I drop the `Tag` suffix, and use an alias for `var`?

machour commented:
ok, perfect!

machour commented:
Renamed classes and added the following function in src/Tag.php, used both internally and in [`src/tag/Columns.php`](https://github.com/mpdf/mpdf/pull/565/files#diff-caa8196509c6f6910999940cf5afa728R9)

```php
	/**
	 * Returns the fully qualified name of the class handling the rendering of the given tag
	 *
	 * @param string $tag The tag name
	 * @return string The fully qualified name
	 */
	public static function getTagClassName($tag)
	{
		static $map = [
			'COLUMN_BREAK' => 'Columnbreak',
			'PAGE_BREAK' => 'Pagebreak',
			'VAR' => 'VarTag',
		];

		$className = 'Mpdf\Tag\\';
		$className .= isset($map[$tag]) ? $map[$tag] : ucfirst(strtolower($tag));

		return $className;
	}
```

Do we want to rename more classes, like change `Sethtmlpagefooter` to `SetHtmlPageFooter`, or is it okay for now?

Also, since this is a massive change, and everything isn't currently covered by unit testing, I was wondering if there is a reference mpdf generated file using all HTML tags, that could serve as a reference to validate that there are no breakage ?

I would happily take care of it if not :)

finwe commented:
If you don't mind, SetHtmlPageFooter certainly looks better.

For testing, I generally use the [examples repository][1] and run the new code on top of the files and then compare output and file sizes to reference files.

[1]:https://github.com/mpdf/mpdf-examples

machour commented:
Thank you, I'll run samples to check all that.

Here's the proposed final renaming. Could you edit this comment and suggest alternatives if needed?

You'll notice that small classes sometimes gets all uppercased, and sometimes not. The reasoning is that `Em` is for _em_phasis so only E is uppercased, while `TD` is Table Data, so full uppercase.

|Tag|Class name|Fix?|
|----|------------|----|
|A|A| <!-- fix here --> |
|ACRONYM|Acronym| <!-- fix here --> |
|ADDRESS|Address| <!-- fix here --> |
|ANNOTATION|Annotation| <!-- fix here --> |
|ARTICLE|Article| <!-- fix here --> |
|ASIDE|Aside| <!-- fix here --> |
|B|B| <!-- fix here --> |
|BARCODE|BarCode| <!-- fix here --> |
|BDI|BDI| <!-- fix here --> |
|BDO|BDO| <!-- fix here --> |
|BIG|Big| <!-- fix here --> |
|BLOCKQUOTE|Blockquote| <!-- fix here --> |
|BOOKMARK|Bookmark| <!-- fix here --> |
|BR|BR| <!-- fix here --> |
|CAPTION|Caption| <!-- fix here --> |
|CENTER|Center| <!-- fix here --> |
|CITE|Cite| <!-- fix here --> |
|CODE|Code| <!-- fix here --> |
|COLUMNBREAK|ColumnBreak| <!-- fix here --> |
|COLUMNS|Columns| <!-- fix here --> |
|DD|DD| <!-- fix here --> |
|DEL|Del| <!-- fix here --> |
|DETAILS|Details| <!-- fix here --> |
|DIV|Div| <!-- fix here --> |
|DL|Dl| <!-- fix here --> |
|DOTTAB|DotTab| <!-- fix here --> |
|DT|DT| <!-- fix here --> |
|EM|Em| <!-- fix here --> |
|FIELDSET|FieldSet| <!-- fix here --> |
|FIGCAPTION|FigCaption| <!-- fix here --> |
|FIGURE|Figure| <!-- fix here --> |
|FONT|Font| <!-- fix here --> |
|FOOTER|Footer| <!-- fix here --> |
|FORM|Form| <!-- fix here --> |
|FORMFEED|FormFeed| <!-- fix here --> |
|H1|H1| <!-- fix here --> |
|H2|H2| <!-- fix here --> |
|H3|H3| <!-- fix here --> |
|H4|H4| <!-- fix here --> |
|H5|H5| <!-- fix here --> |
|H6|H6| <!-- fix here --> |
|HEADER|Header| <!-- fix here --> |
|HGROUP|HGroup| <!-- fix here --> |
|HR|HR| <!-- fix here --> |
|I|I| <!-- fix here --> |
|IMG|Img| <!-- fix here --> |
|INDEXENTRY|IndexEntry| <!-- fix here --> |
|INDEXINSERT|IndexInsert| <!-- fix here --> |
|INPUT|Input| <!-- fix here --> |
|INS|Ins| <!-- fix here --> |
|KBD|Kbd| <!-- fix here --> |
|LEGEND|Legend| <!-- fix here --> |
|LI|LI| <!-- fix here --> |
|MAIN|Main| <!-- fix here --> |
|MARK|Mark| <!-- fix here --> |
|METER|Meter| <!-- fix here --> |
|NAV|Nav| <!-- fix here --> |
|NEWCOLUMN|NewColumn| <!-- fix here --> |
|NEWPAGE|NewPage| <!-- fix here --> |
|OL|OL| <!-- fix here --> |
|OPTION|Option| <!-- fix here --> |
|P|P| <!-- fix here --> |
|PAGEBREAK|PageBreak| <!-- fix here --> |
|PAGEFOOTER|PageFooter| <!-- fix here --> |
|PAGEHEADER|PageHeader| <!-- fix here --> |
|PRE|Pre| <!-- fix here --> |
|PROGRESS|Progress| <!-- fix here --> |
|Q|Q| <!-- fix here --> |
|S|S| <!-- fix here --> |
|SAMP|Samp| <!-- fix here --> |
|SECTION|Section| <!-- fix here --> |
|SELECT|Select| <!-- fix here --> |
|SETHTMLPAGEFOOTER|SetHtmlPageFooter| <!-- fix here --> |
|SETHTMLPAGEHEADER|SetHtmlPageHeader| <!-- fix here --> |
|SETPAGEFOOTER|SetPageFooter| <!-- fix here --> |
|SETPAGEHEADER|SetPageHeader| <!-- fix here --> |
|SMALL|Small| <!-- fix here --> |
|SPAN|Span| <!-- fix here --> |
|STRIKE|Strike| <!-- fix here --> |
|STRONG|Strong| <!-- fix here --> |
|SUB|Sub| <!-- fix here --> |
|SUMMARY|Summary| <!-- fix here --> |
|SUP|Sup| <!-- fix here --> |
|TABLE|Table| <!-- fix here --> |
|TBODY|TBody| <!-- fix here --> |
|TD|TD| <!-- fix here --> |
|TEXTAREA|TextArea| <!-- fix here --> |
|TEXTCIRCLE|TextCircle| <!-- fix here --> |
|TFOOT|TFoot| <!-- fix here --> |
|TH|TH| <!-- fix here --> |
|THEAD|THead| <!-- fix here --> |
|TIME|Time| <!-- fix here --> |
|TOC|Toc| <!-- fix here --> |
|TOCENTRY|TocEntry| <!-- fix here --> |
|TOCPAGEBREAK|TocPageBreak| <!-- fix here --> |
|TR|TR| <!-- fix here --> |
|TT|TT| <!-- fix here --> |
|TTA|TTA| <!-- fix here --> |
|TTS|TTS| <!-- fix here --> |
|TTZ|TTZ| <!-- fix here --> |
|U|U| <!-- fix here --> |
|UL|UL| <!-- fix here --> |
|VAR|VarTag| <!-- fix here --> |
|WATERMARKIMAGE|WatermarkImage| <!-- fix here --> |
|WATERMARKTEXT|WatermarkText| <!-- fix here --> |

finwe commented:
Done.

finwe commented on src/Tag.php:
```diff
@@ -130,9 +130,37 @@ private function getTagInstance($tag)
 	public static function getTagClassName($tag)
```
Is there a need for the function to be static? And public?

machour replied:
Used in Columns.php

Klap-in replied:
Is than protected still useful?

finwe replied:
No, not really, when paradigm shifted from inheritance to composition. Can be handy in some cases, but common best practice is to avoid it. (If I understand your somewhat off-topic question correctly).

Klap-in replied:
sorry, for these not-expert (but still-interested) questions/remarks... mainly for my understanding about your design decisions...
I mean, if you like to limit the scope of the getTagClassName function, you could use it as protected static function. Because it is only used in Tag and inherited tags. 
A function that is protected can be only used here, so if you like to shift to another approach, you have only these classes that could actually use it. Therefore you know that you will break nothing by removing it at that time.

machour commented:
@Klap-in sec/tag/Columns.php doesn't inherit from src/Tag.php ;)

Klap-in commented:
Ah.. namespaces, thanks for the patience :-)

finwe commented:
Errors and notices encountered when running the last version on example files (unique):

```
PHP Notice:  Undefined index: In in /home/finwe/www/_github/mpdf/repo/src/Tag/Input.php on line 423
PHP Warning:  A non-numeric value encountered in /home/finwe/www/_github/mpdf/repo/src/Mpdf.php on line 21762
PHP Warning:  A non-numeric value encountered in /home/finwe/www/_github/mpdf/repo/src/Mpdf.php on line 21765
PHP Warning:  A non-numeric value encountered in /home/finwe/www/_github/mpdf/repo/src/Mpdf.php on line 24142
PHP Notice:  Undefined index: In in /home/finwe/www/_github/mpdf/repo/src/Tag/Input.php on line 119
PHP Notice:  Undefined index: FOOTER-STYLE in /home/finwe/www/_github/mpdf/repo/src/Tag/PageFooter.php on line 42
PHP Notice:  Undefined index: FOOTER-STYLE-RIGHT in /home/finwe/www/_github/mpdf/repo/src/Tag/PageFooter.php on line 118
PHP Notice:  Undefined index: FOOTER-STYLE-LEFT in /home/finwe/www/_github/mpdf/repo/src/Tag/PageFooter.php on line 74
PHP Warning:  A non-numeric value encountered in /home/finwe/www/_github/mpdf/repo/src/Mpdf.php on line 13635
PHP Notice:  Undefined offset: 0 in /home/finwe/www/_github/mpdf/repo/src/Mpdf.php on line 26622
PHP Fatal error:  Uncaught Mpdf\MpdfException: Error detected. PDF file generation aborted: Undefined offset: 0 in /home/finwe/www/_github/mpdf/repo/src/Mpdf.php:26622
Stack trace:
#0 /home/finwe/www/_github/mpdf/repo-examples/example26_RTL.php(358): Mpdf\Mpdf->Output()
#1 {main}
  thrown in /home/finwe/www/_github/mpdf/repo/src/Mpdf.php on line 26622
PHP Notice:  Undefined index: x in /home/finwe/www/_github/mpdf/repo/src/Tag/Meter.php on line 287
PHP Notice:  Undefined index: y in /home/finwe/www/_github/mpdf/repo/src/Tag/Meter.php on line 288
```

machour commented:
Hi @finwe and merry Christmas! 🎄 

I fixed all problems, they were all cause by a really wrong tag name computation: https://github.com/mpdf/mpdf/pull/565/commits/7da979945d28eefaa1b266920bc3b82ae14340ce

I ran all examples with current `development` branch and mine, and everything seems okay.

I also found a cool trick to tests for differences by using `convert` to transform PDF files into JPEG.
md5sums are pretty much the same, expect for PDF generated using randomized data. Might be something worth considering for `mpdf-examples` ;)

finwe commented:
Hey! It looks fine, thanks! Could you please rebase to development instead of creating the merge commit? There is a one commit altering Tag.php, but resolving the conflicts should not be too hard, it is only one or two lines changed.

machour commented:
@finwe I reverted the merge commit and tried to rebase, but this is my first time ever doing this and I'm under the impression that I messed it up :/
Could you check and tell me?

machour commented:
Here's what I did:

```
# reverted merge commit
git revert 4dc445f62c7b988adc3f836bb2fd22cbe2e84325 -m 1
# rebased
git rebase development
# fixed conflicts and applied changes manually then
git rebase --continue
# git push failed, so 
git pull
# fixed a whitespace conflict then
git add src/Tag.php
git push
```

finwe commented:
I gave it a go in `tag-refactoring` branch - can you please verify everyting is OK?

machour commented:
Thank you so much. Ran the examples on the new branch, everything seems fine!

finwe commented:
Thanks!
